### ISSUE TITLE
Improve quoted shell parameter handling on Windows

### ISSUE THREAD

**Issue by chrisbouchard** [alacritty/alacritty#3552]
When launching Alacritty with a custom command via `alacritty.exe -e`, it seems that Alacritty does not preserve the quoting of its arguments. E.g.,

```
alacritty.exe -e wsl zsh -c 'cd ~ && pwd && zsh -i'
```

On my system, this opens a new window and prints `/mnt/c/Users/cbouchard`, my Windows home directory, which is the current directory of the shell running the command. So seemingly the `cd` did nothing. However,

```
alacritty.exe -e wsl zsh -c "'cd ~ && pwd && zsh -i'"
```

(Note the double-quoting.) This opens an new window and prints `/home/cbouchard`, which is my WSL home directory, i.e., `~`. So the `cd` worked as desired.

My best guess is that Alacritty is dropping the quoting from my positional parameters, so it's interpreting the first command as

```
wsl zsh -c cd ~ && pwd && zsh -i
```

So it runs the `cd` in a subshell, then runs `pwd` and `zsh -i`.

I feel like, if Alacritty will accept positional parameters as separate parameters, it should preserve the quoting/spaces of those parameters.

### System

OS: Windows
Version: alacritty 0.4.2 (f68de37)
Linux/BSD: X11/Wayland, DE or Compositor and WM
Windows: ConPTY

**Comment by chrisduerr** [comment]
Alacritty does indeed not pass quotes to the shell program executed, if they are 'top level' quotes.

This is simply because the command parser interprets them as arguments to Alacritty and thus assumes that these quotes were used to make sure the parameters are correctly passed to Alacritty.

That's why if there's a desire to pass quotes to the shell, you need to quote it to tell Alacritty's command parser what is 'one parameter' and then need to add quotes again to make it clear that this single parameter contains quotes.

I'd say fundamentally that's in line with how most software works. The top level quotes usually aren't even seen by the application, so there's not much to do about that. Even if we would want to read the quotes by parsing CLI arguments ourselves (which we don't), we couldn't get the quote information you're asking for from `std::env::args()`. They're filtered out before ever reaching our application.

So I suppose fundamentally I'd just advise to quote everything you pass to `zsh -c` if there are any troubles, just how you've already figured out.

@kchibisov @nixpulvis do you think we should change anything here?

**Comment by kchibisov** [comment]
>we couldn't get the quote information you're asking for from std::env::args(). They're filtered out before ever reaching our application.

>@kchibisov @nixpulvis do you think we should change anything here?

If we don't know anything about quotes then we can't do anything?

**Comment by chrisbouchard** [comment]
I find this answer very confusing. I hope I can explain why.

>Alacritty does indeed not pass quotes to the shell program executed, if they are 'top level' quotes.
>
> This is simply because the command parser interprets them as arguments to Alacritty and thus assumes that these quotes were used to make sure the parameters are correctly passed to Alacritty.

If this is true, what you are saying is there is no effective difference between `alacritty -e echo a b c` and `alacritty -e echo "a   b   c"`. They will both result in `echo a b c` (no quotes) being run by Alacritty. But this does not match up with my experience with argument parsing. Surely if I quoted something _to_ alacritty, it's because I intended it to be passed as a single string to the underlying `exec` call.

> I'd say fundamentally that's in line with how most software works. The top level quotes usually aren't even seen by the application, so there's not much to do about that. Even if we would want to read the quotes by parsing CLI arguments ourselves (which we don't), we couldn't get the quote information you're asking for from std::env::args(). They're filtered out before ever reaching our application.

I this may be a misunderstanding. When I said Alacritty should "preserve the quoting", I meant in the bash scripting sense &mdash; it should pass strings as it received them. So yes, I agree that the literal quotes are removed by the shell before it runs the command, but it's not true that there "no way to get the quote information". In my example above, in the first case `env::args()` should look like `[..., "-e", "echo", "a", "b", "c"]`, and in the latter case it should look like `[..., "-e", "echo", "a   b   c"]`. So I would expect the former to result in a process being invoked with an `argv` of length 4, and in the latter with an `argv` of length 2.

And just as an example, this is how Gnome Terminal works. When I run

```
gnome-terminal -- zsh -c 'cd ~ && pwd && zsh -i'
```

I see `/home/chris` as expected (I ran that on my Linux laptop because I don't have my Windows laptop booted up yet this morning.) In this case, `--` is Gnome Terminal's equivalent of Alacritty's `-e`.

**Comment by chrisduerr** [comment]
> So I would expect the former to result in an echo process being invoked with an argv of length 4, and in the latter with an argv of length 2.

And this is exactly what Alacritty does at the moment.

> And just as an example, this is how Gnome Terminal works.

This works because gnome terminal always spawns a shell afaik.

**Comment by nixpulvis** [comment]
@chrisbouchard your example works perfectly for me:

```sh
alacritty -e zsh -c 'cd ~ && pwd && zsh -i'
```

I end up in `zsh` with my home directory printed in the first like of the newly spawned alacritty instance.

**Comment by chrisduerr** [comment]
That does of course work because there's only one shell involved.

An example of what doesn't would be this:

```
alacritty -e zsh -c zsh -c 'cd ~ && pwd && zsh -i'
```

That's a little more like the original `wsl` command posted.

**Comment by chrisbouchard** [comment]
Could this be a bug in the Windows build? I don't think it's the multiple shells, because WSL also forwards its arguments. I've attached a couple screenshots with what I'm seeing.

In the first screenshot, I run the `wsl` command through PowerShell. WSL keeps the ZSH command as a single argument when invoking the shell. Then I run the same command using `alacritty.exe -e`.
![2020-04-06 Alacritty 01](https://user-images.githubusercontent.com/4512028/78578157-1884a980-77fd-11ea-849b-e61396807059.png)

The result is this second Alacritty window.
![2020-04-06 Alacritty 02](https://user-images.githubusercontent.com/4512028/78578161-1ae70380-77fd-11ea-8669-e43f2d052e9b.png)

Note: Your example with two layers of ZSH isn't quite accurate, because `zsh -c` expects to receive the command as a _single_ string argument. WSL works like Alacritty, accepting the arguments as separate arguments.

**Comment by nixpulvis** [comment]
Just to add to this, and possibly this clears things up @chrisbouchard? We can use [`$@`](https://www.tldp.org/LDP/abs/html/internalvariables.html#APPREF) to print the command arguments as is (periods added to help see variable expansion).

The `man` page for `sh` gives us an idea of how this will be interpreted:

    command_string [command_name [argument...]]

Here we pass `foo` as `command_name` (aka `$0`) and `1 2 3 4` as `argument...` (aka `$@`, or `$*`) to the **subshell** (this is the key part), which are then used in the expansion before finally being printed by `echo`.

```
$ sh -c 'echo .$0. .$@.' foo 1 2 3 4
.foo. .1 2 3 4.
```

Noting that we used single quotes in the above example, if we try the same thing with double quotes we can see that they are expanded **before** being passed to echo, and therefor `$@` is empty, since my top level `sh` doesn't have arguments.

```
$ sh -c "echo .$0. .$@." foo 1 2 3 4
.sh. ..
```

However, if we quote the whole command (regardless of the type of quote), `echo` is given arguments **after** the subshell has expanded `$@` to the empty string (since we didn't pass it any arguments after the `command_string`), and `$0` is simply `sh` since that is command being run at the time of expansion here.

```
$ sh -c 'echo .$0. .$@. foo 1 2 3 4'
.sh. .. foo 1 2 3 4

# The printed "sh" below would be whatever shell you're running all this
# within, in this case everything is within "sh".

$ sh -c "echo .$0. .$@. foo 1 2 3 4"
.sh. .. foo 1 2 3 4

```

Finally, we can parse a nested shell, like the one @chrisduerr posted above. The first `sh -c` works as always, and takes the following `sh` and executes it. Now however, the following `-c ...` is an **argument** to the **first** subshell, which has been instructed to run the second `sh` without any further arguments.

```
$ sh -c sh -c 'echo .$0. .$@. foo 1 2 3 4'
# prints nothing, now running a nested `sh`
$ exit
```

I hope this helped. Took me a little playing around to understand myself. Nested shells and escapes are always a little tricky.

**Comment by nixpulvis** [comment]
@chrisbouchard here's what Alacritty is seeing with and without two layers of quotes, as in your original example:

**Only single quotes:**
```
...
```

**With single quotes within double quotes:**
```
...
```

**Comment by chrisduerr** [comment]
> Could this be a bug in the Windows build? I don't think it's the multiple shells, because WSL also forwards its arguments. I've attached a couple screenshots with what I'm seeing.

Yes. On Windows we just collect everything into a big String and pass that to their "PTY".

Since Windows unfortunately just takes a big String (see [here](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw)).

So we just to `program + args.join(" ")` basically it seems. If someone has a Windows machine it should be possible to test wrapping arguments in quotes again, but that can likely cause problems. So fundamentally Windows just sucks here. But if you're willing to look into it you could probably improve it a bit.

**Comment by chrisbouchard** [comment]
@chrisduerr: That does explain what I'm seeing, thank you. It's nothing to do with shell quoting, it's just that the Windows version doesn't follow the same rules.

I would be interested to look into improving that &mdash; I've never worked that low level on Windows but I've been enjoying using Alacritty for work. There _must_ be a way, because `cmd` et. al. are able to start processes involving arguments with spaces in their names (e.g., filenames). Maybe tonight I can look at how Alacritty starts the subprocess and try adding some quotes to see what happens. (And if that does work, it should probably make sure to escape any quotes in the input arguments as well.)

**Comment by nixpulvis** [comment]
@chrisbouchard yea sorry, I didn't realize the issue until I tried it out on Windows for myself.

FWIW, here's where the code lives: https://github.com/alacritty/alacritty/blob/13eb50de799c4c6f7ebed1a7fa1ab25b378580bb/alacritty_terminal/src/tty/windows/mod.rs#L305-L313

**Comment by chrisbouchard** [comment]
Hoo boy, found https://docs.microsoft.com/en-us/cpp/c-language/parsing-c-command-line-arguments?view=vs-2019, which lays out the quoting rules. Adding the quotes shouldn't be a problem, but escaping the arguments will require some care.

**Comment by chrisbouchard** [comment]
Here's an attempt at implementing the quoting rules: https://github.com/chrisbouchard/alacritty/blob/quote-windows-args/alacritty_terminal/src/tty/windows/mod.rs#L316-L380

I haven't had a chance to actually try this on Windows yet. (I at least know it compiles.) Once I have, I can add some tests and create a pull request.

**Comment by chrisduerr** [comment]
@chrisbouchard Note that you wouldn't want to take a String as input.

We already have the args split, just like on Linux. The thing you'd need to do is surround each parameter with quotes, to separate them.

The problem with it is just that the quotes might also be part of the parameter already, so you need to make sure those are escaped appropriately. And if there already are escaped quotes, you need to make sure the escape is escaped appropriately.

**Comment by chrisbouchard** [comment]
@chrisduerr Yes, thank you. I believe the function I linked last night does what you describe. I just had a chance to try it out on my Windows laptop, and it correctly runs the `wsl` command in the issue description.

**Comment by chrisbouchard** [comment]
Ok, it's definitely better &mdash; I'm able to pass arguments containing whitespace and have them come across as a single argument to the command. But I'm only getting more puzzled as I try to test escaping existing quotes in the arguments. I was getting tired of trying to figure out what `wsl` or other programs were doing with the command line, so I wrote a very simple Rust program:

```rust
use std::{env, thread, time};

fn main() {
    println!("{:?}", env::args());
    thread::sleep(time::Duration::from_secs(30));
}
```

(The sleep is just there so I can use it as the command for Alacritty and see the output before the window closes.)

Now check out this PowerShell session:

```
PS C:\Users\cbouchard\External\alacritty> echo 'abc "def ghi"'
abc "def ghi"
PS C:\Users\cbouchard\External\alacritty> argv-echo.exe 'abc "def ghi"'
Args { inner: ["C:\\Users\\cbouchard\\bin\\argv-echo.exe", "abc def", "ghi"] }
```

Here `echo` is a PowerShell builtin, so you can see that PowerShell kept the inner double-quotes as part of the argument. But, when I call my simple program, not only are the quotes gone (the quotes you see there are just Rust's debug format for strings) but the single argument has been broken into two at the _second_ space.

Continuing,

```
PS C:\Users\cbouchard\External\alacritty> echo 'abc \"def ghi"'
abc \"def ghi"
PS C:\Users\cbouchard\External\alacritty> argv-echo.exe 'abc \"def ghi"'
Args { inner: ["C:\\Users\\cbouchard\\bin\\argv-echo.exe", "abc \"def ghi\""] }
```

Now I've added an extra backslash to the argument (which you can see in the `echo` output is not interpreted as an escape &mdash; Powershell does not use C-style escapes), and now in my program's output the argument is _not_ split, and _both_ quotes are included (here the backslashes _are_ an escape, again due to Rust's debug format for strings).

I really don't understand what's going on here exactly. I think my next step will be to rewrite my simple program in C and see if I see the same behavior. I just don't understand if this is _expected_ behavior for PowerShell (if it is, it's insane), or if there's something wrong deeper down even than Alacritty.

---

tl;dr: My patch makes it better, but something's still weird.

**Comment by chrisbouchard** [comment]
Oh FFS... It's _PowerShell_ that's "wrong"!

https://stackoverflow.com/questions/6714165/powershell-stripping-double-quotes-from-command-line-arguments

**Comment by chrisbouchard** [comment]
Ok, I'll leave this up to you guys. I have a fix that seems to implement the rules described in [Everyone quotes command line arguments the wrong way](https://docs.microsoft.com/en-us/archive/blogs/twistylittlepassagesallalike/everyone-quotes-command-line-arguments-the-wrong-way) (the function I wrote is very similar to the one they describe). It's worth reading to get a handle on this mess. (I'm not a fan of the tone &mdash; Microsoft seems to be selling that these quoting rules are good &mdash; but the information is good.)

However, as they mention in that article, _technically_ we don't know that we _should_ quote and escape the input this way. It's very likely that the program being invoked _does_ use `CommandLineToArgvW` to parse the command line, in which case we'd want to do as we've been discussing. But it's possible the command is something like `cmd /c` that would interpret the command line literally &mdash; in that case quoting and escaping could make things worse!

I suppose these are the options I see:

1. Leave it as-is, and add documentation describing what Alacritty does for Windows commands and why, and accept that it will be annoying\* most of the time.
2. Implement the quoting and escaping, and accept that it will be annoying for people who _do_ want to run `cmd /c`.
3. Add a flag to disable quoting and escaping for a subsequent `-e`.
4. Add an alternate version of `-e` that does not do quoting and escaping.

\*: It's more than just annoying, because it requires more than just double-quoting. It would actually require users to know and use the ecaping rules described in that article I linked above for their arguments. But PowerShell makes people do that too, so it's not unprecedented.

**Comment by chrisduerr** [comment]
> I believe the function I linked last night does what you describe.

It does not. You're trying to solve the problem of splitting a Windows commandline into arguments. We'd need to solve the problem of maintaining argument separation when turning an array of arguments into a commandline.

> Ok, I'll leave this up to you guys.

I personally don't have much interest at all to waste my time working around awful Windows APIs and that's likely the same for everyone else currently part of the Alacritty team.

So this is probably not going to change unless someone who actually uses Windows is interested in resolving it.

**Comment by chrisbouchard** [comment]
> It does not. You're trying to solve the problem of splitting a Windows commandline into arguments. We'd need to solve the problem of maintaining argument separation when turning an array of arguments into a commandline.

I'm getting quite frustrated with this. I am _not_ re-splitting the argument list. This function is looping over a single argument. I have to go character by character to implement Windows' context sensitive backslash-escaping rules. Please stop assuming I'm some sort of idiot.

> I personally don't have much interest at all to waste my time working around awful Windows APIs and that's likely the same for everyone else currently part of the Alacritty team.

I agree, Windows' API here is awful. But Winows isn't exactly a minority platform. I'm offering to do the work, so I don't get why you're being so dismissive of a request for a direction.

> So this is probably not going to change unless someone who actually uses Windows is interested in resolving it.

Yes, exactly. So how about at least giving me the time of day so I can _do_ that?

**Comment by nixpulvis** [comment]
@chrisbouchard FWIW, I was interested in testing this out, I just haven't had time yet. There's no need to rush this feature through.

Sorry if I contributed to the tone of this thread, I really appreciate the work you've put into investigating this issue.

**Comment by chrisduerr** [comment]
> I'm getting quite frustrated with this. I am not re-splitting the argument list. This function is looping over a single argument.

Fair enough. I thought you were suggesting to pass the entire commandline to that function.

> But Winows isn't exactly a minority platform.

As far as Alacritty's support goes, it is. It's definitely in a state of reduced support and less importance. We've added it since it has been graciously provided by contributors, but I don't think there's any desire to put in as much work as we do for other platforms at least from the core contributors.

I personally don't really care how many users a specific platform has. But even if that's your argument, it should be considered that the number of terminal users on macOS/Linux is significantly higher than on Windows. And Alacritty's Windows users are significantly lower than the users of the other platforms as far as I am aware.

> I'm offering to do the work, so I don't get why you're being so dismissive of a request for a direction.

You said the following:

> Ok, I'll leave this up to you guys.

Which goes in a bit of a different direction than doing the work. If you're willing to see this through, please create a pull request.

**Comment by chrisbouchard** [comment]
Hello again. I'd like to revisit this. I cleaned up and rebased my branch and opened a merge request: #3857.

I apologize if I did not communicate clearly; when I said &ldquo;I'll leave this up to you guys,&rdquo; I was referring to a decision to be made regarding the concern I raised in that comment. Escaping is _not_ always the right thing to do, but it's _often_ the right thing to do. Unfortunately, as far as I can tell, there's no way to tell whether escaping in this manner is safe without knowing how the shell we're calling was written &mdash; i.e., whether it calls `CommandLineToArgvW` like a C/C++/Rust program, or uses custom command line parsing like `cmd`. I'd hate to solve my use case at the expense of other users.

I have added a check to avoid escaping arguments when possible, so that should make always escaping a little safer.

**Comment by jeot** [comment]
Hello @chrisbouchard 
I had a problem that I think may be related to the issue you have mentioned here. So I thought I would share...
I'm a Windows and Neovim user. After trying lots of terminals and GUI implementation for Neovim, I have reached the conclusion that Alacritty is my best choice for running Neovim. (Thanks Alacrity!). Here is the problem...
If I run the following command in Alacritty (or any other terminal like Windows Terminal) it will perform without any problem:
```
nvim "C:\Users\shk\Downloads\Labs\Hands-on_1-1 G0_App\G0_App\Src\app.c"
```
It opens the `app.c` file correctly in the given path.

But if I run the following command (in Alacrity or Windows Terminal):
```
alacritty -e nvim "C:\Users\shk\Downloads\Labs\Hands-on_1-1 G0_App\G0_App\Src\app.c"
```
It will simply ignore the quotes and don't open the actual file, but instead opens two empty files with default names of `Hands-on_1-1` and `app.c` as shown in the picture below:
<img width="420" alt="image" src="https://github.com/alacritty/alacritty/assets/4253298/be433baa-8ebc-4c39-9170-27fa454ef941">
The main reason that I stumbled upon this problem was that I was trying to add a new menu when right-clicking a file or directory to open with Neovim inside Alacrity. I also tried your first hack which was simply adding an extra single colon, but it didn't work.
I also saw your work (pull request #3857) and don't understand why it was not even considered! Also, I'm too noob to even fork your code and try to build for myself. An amateur programmer using Alacritty on Windows in 2023!
I'm just wondering how Alacrity can simply run the first command and mess up the second command, both running from inside its own terminal!

**Comment by osamasalem** [comment]
same issue here with NeoVim under Windows

**Comment by chrisduerr** [comment]
Closing since I think manually escaping is the better default, since it allows doing everything without requiring an additional config option.
